### ISSUE TITLE
Analyze www/no-www dataset

### ISSUE THREAD

**Issue by gbinal** [GSA/site-scanning#1044]
Once #1017 is complete, we should:  

- host a copy [here](https://github.com/GSA/site-scanning-snapshots/tree/main/other/misc)
- Pull it into Sheets and ask:  
  - Count:
    - final_url - 
    - final_url_status_code -
    - primary_scan_status - 
    - www_final_url - 
    - www_status_code - 
    - www_scan_status - 
    - www_same - 
  - Analyze deltas of: 
    - final URLs - 
    - status codes - 
    - scan statuses - 
  - What else?  

Questions to ask:  

- What is the expected or desired behavior and how often is that not the case?
 - when target url and target domain are the same, and one works, the other should, too.  (or, there's one period in the target URL)
   - a bad practice is if one works and the other doesn't.  
   - another bad practice is if the final URLs are different 
 - when there's a subdomain (different target URL and target URL domain OR when there's 2+ periods in the target URL)...
   - if the naked target URL worked, then adding www. before wouldn't work.  
   - note any that that have working with www and not working naked, both to call out, but also to highlight as websites that we'll lose the data for in future scans.  
   - note any that both with and without both work. 
- Can/should we switch to a target URL list without www. at the beginning and what would be the impact of that?
- Where are there anti patterns that should be called out?
  - use of www3 etc.

**Comment by gbinal** [comment]
Doing this [here](https://docs.google.com/spreadsheets/d/1cjbyNdia3I3Vgjt8IfixDu0tsSDM0Sdrua3DzUc6sXo/edit#gid=2665449).

- Out of 26,164 - 

Primary Scan Status | Count
-- | --
timeout | 3039
dns_resolution_error | 6437
completed | 15770
unknown_error | 129
connection_refused | 209
execution_context_destroyed | 123
connection_reset | 133
connection_closed | 120
evaluation_failed | 10
too_many_redirects | 7
address_unreachable | 57
ssl_version_cipher_mismatch | 56
invalid_ssl_cert | 31
empty_response | 23
http2_error | 19
 


www Scan Status | Count
-- | --
dns_resolution_error | 20623
timeout | 453
completed | 4554
http2_error | 68
unknown_error | 40
ssl_version_cipher_mismatch | 45
connection_reset | 26
invalid_ssl_cert | 1
address_unreachable | 9
connection_refused | 10
connection_closed | 7
empty_response | 2
(blank) | 325

**Comment by abbybowman** [comment]
Hey @gbinal do you have any time today or tomorrow to hop on a call about this issue? I was up late last night trying to debug some issues with our own internal domain tracker (that pulls in Site Scanning data) and it turns out to originate in this exact issue. I have 5 example domains from NASA I looked at where Site Scanning has separate rows for www and non-www domains, and went through each column to mark which row (www or non) had the correct results. 

TLDR, it's not consistent, and I wanted to brainstorm with you on how to get an accurate consolidated scan result for the sites where www redirects to non-www or vice versa (and what do we do when best practices aren't followed, like when the bare domain doesn't redirect to the live www version, e.g. grc.nasa.gov).

I came into the issue list and was thrilled to see this here and that you and the team are already looking into this. It's a big blocker for us right now, so would greatly appreciate thinking through this with you (esp if there are any quick fixes we can put into place on our side with data processing).

**Comment by gbinal** [comment]
Note to self - I should make a point to upload this dataset after for later review.  

Also! - include a snapshot of the last modified header dataset

**Comment by gbinal** [comment]
note - 325 www scans appeared to have not run 


Alright - took a fresh pass at this and here's what I'm seeing.  

- Of 26163 target URLs, 1437 have no subdomain (only one period in the target URL).  
  - 1176 of the 1437 have the same status code
  - 261 of the 1437 then have different status codes.  
    - 
    - This is bad.  These sites 


Take 2: 

- Removing the 325 rows that for some reason had no www-scan results, there are 25838 target URLs.  
- Of those 25838, 1409 have no subdomain (only one period in the target URL).
  - 1162 of the 1409 have the same status code.  
  - 247 of the 1409 have different status codes.  
    - This is bad.  For any x.gov, it should work with or without the www.  
    - 72 work without a www., but not with.  
    - 162 work with a www., but not without. 
  - There's 809 that return 200/202 status codes both with and without www.
    - 134 have identical resulting final URLs.  
    - 675 have different resulting final URLs
      - This is not optimal.  
      - In particular, 139 go to different pages, which is not good.  In most cases, it's a matter that x.gov resolves as x.gov and www.x.gov resolves as www.x.gov.  I'm not sure if that's suboptimal or not.  
      - Virtually all of the others are a matter of one ending  with a slash and the other not.  I'm not sure if that is suboptimal or not.  
  - Note that 544 of the 1162 redirect no www to www. and 101 redirect www to no www

Of the 24,429 target URLs that have a subdomain (2+ periods), 10,331 have the same status code.  

- Of those 10,331, 1576 return 200 status codes both with and without www.  
- The other 8755 fail both with and without www.  
- Of the 14,098 that have different results depending on if there's a www, 10,466 work without a www but not when one is added and,
  - 1265 work with a www. but not without.  
  - 2367 fail in different ways 

also compare scan statuses 
also compare final URLs for ones with same status code 




What is the expected or desired behavior and how often is that not the case?
when target url and target domain are the same, and one works, the other should, too. (or, there's one period in the target URL)
a bad practice is if one works and the other doesn't.
another bad practice is if the final URLs are different
when there's a subdomain (different target URL and target URL domain OR when there's 2+ periods in the target URL)...
if the naked target URL worked, then adding [www](http://www/). before wouldn't work.
note any that that have working with www and not working naked, both to call out, but also to highlight as websites that we'll lose the data for in future scans.
note any that both with and without both work.
Can/should we switch to a target URL list without [www](http://www/). at the beginning and what would be the impact of that?
Where are there anti patterns that should be called out?
use of www3 etc.

**Comment by gbinal** [comment]
Wrote up ^^^ notes here - https://github.com/GSA/site-scanning-documentation/blob/main/about/project-management/proposals/dropping-www.md
